### DISCUSSION TITLE
IOS Simulator External Display

### DISCUSSION THREAD

**Discussion by pcramb** [dotnet/maui#10684] [Q&A]
I'm trying out MAUI for an IOS app that needs to use screen mirroring to an Apple TV. However, I can't seem to get anything to show up on an external display on the IOS simulator (I'm using a mac). Just wondering if someone could point me in the right direction to get content to the external display?

**Comment by Medeep99** [comment]
Hey, I dont know if this will work for you or not, but i tried it like this and its working correctly right now:

    private UIScreen? _externalScreen;

    private static bool DetectSecondaryDisplayiOS()
    {
        return UIScreen.Screens.Length > 1;
    }
    
  
    private void OpenOnSecondaryDisplayiOS()
    {
        if (_externalScreen == null)
            return;

        var windowScene = UIApplication.SharedApplication.ConnectedScenes
            .OfType<UIWindowScene>()
            .FirstOrDefault(s => s.Screen == _externalScreen);

        if (windowScene == null)
        {
            Task.Delay(300).ContinueWith(_ =>
            {
                MainThread.BeginInvokeOnMainThread(OpenOnSecondaryDisplayiOS);
            });
            return;
        }

        _iosExternalWindow = new UIWindow(windowScene)
        {
            Frame = windowScene.Screen.Bounds,
            BackgroundColor = UIColor.White
        };

        var page = new CustomerDisplayPage(_state);

        // Use the MAUI context already available in the app.
        var mauiContext = Application.Current!.Handler!.MauiContext!;
        var pageController = page.ToUIViewController(mauiContext);

        pageController.View.BackgroundColor = UIColor.White;

        _iosExternalWindow.RootViewController = pageController;
        _iosExternalWindow.Hidden = false;
        _iosExternalWindow.MakeKeyAndVisible();
    }
    private void RegisteriOSScreenNotifications()
    {
        NSNotificationCenter.DefaultCenter.AddObserver(
            UIScreen.DidConnectNotification,
            (NSNotification notification) =>
            {
                _externalScreen = notification.Object as UIScreen;

                MainThread.BeginInvokeOnMainThread(async () =>
                {
                    HasSecondaryDisplay = true;
                    await OpenAsync();
                });
            });

        NSNotificationCenter.DefaultCenter.AddObserver(
            UIScreen.DidDisconnectNotification,
            (NSNotification notification) =>
            {
                MainThread.BeginInvokeOnMainThread(() =>
                {
                    HasSecondaryDisplay = false;
                    CloseIosWindow();
                });
            });
    }

    private void CloseIosWindow()
    {
        if (_iosExternalWindow != null)
        {
            _iosExternalWindow.Hidden = true;
            _iosExternalWindow.RootViewController = null;
            _iosExternalWindow = null;
        }

        _externalScreen = null;
    }


NOTE: The delay is just for my mind's peace. Also, in the simulator it is better to open the external display only after your app runs or loads the page/vm that calls the external page service. That seemed to work for me. I also tried using the notificationCenter to notify if there is an external display and directly pass that display to the _externalDisplay, but it did not work. That wouldve been a much better way of allocating the external display for use by the code.

Please suggest any improvements on the above code if you have bene doing this in another way.
